base/iobuffer.jl

```diff
@@ -59,6 +59,15 @@ It may take optional keyword arguments:
 
 When `data` is not given, the buffer will be both readable and writable by default.
 
+!!! warn "`data` is invalid after being passed to `IOBuffer`"
+    Once `data` is passed to `IOBuffer`, if `write=true` (or `append=true`) and
+    `maxsize > length(data)`, `IOBuffer` essentially "owns" this data, meaning that
+    existing bindings to `data` should be considered invalid. There are no guarantees on
+    its content at this point. For example, `IOBuffer` may re-allocate the data as required,
+    which may or may not be visible in any outstanding bindings to `array`. There could be
+    unallocated, arbitrary values. Likewise, the ordering of any values written to this
+    array cannot be presumed.
```

Comment:
In addition to no guarantees on its content, future mutation of `data` can cause unexpected behavior including data corruption.

By documenting it this way, we guarantee that if write and append are false or maxsize is equal to length(data) then `data` is valid after being passed. I want to make sure that API choice is intentional.

Reply:
That seems slightly more an issue with the `String` constructor (and continuing to use the argument afterwards) in v1.11 than with `IOBuffer`. In v1.12, that code is now more safe (due to Memory) and won't segfault (though mutating the Array/Memory after calling String is currently still invalid)

Reply:
Should calling `IOBuffer` on a `Vector{UInt8}` with `write=true` empty the original vector after taking the memory, like the `String` function currently does?

Reply:
Sure, if `write=true` and `maxsize > length` that seems a good way to partially encode this restriction in code